### ISSUE TITLE
API Bug

### ISSUE THREAD

**Issue by jmercouris** [atlas-engineer/nyxt#2782]
**Describe the bug**

I'm really excited about our new adjustable column prompt buffer. It is awesome. That said, there is still room for improvement! I think that the following is an API error

```
(define-class test-source (prompter:source)
  ((prompter:name "Test")
   (prompter:constructor '("suggestion"))))

(defmethod prompter:object-attributes ((object string) (source test-source))
  `(("Attribute-1" ,object nil)
    ("Attribute-2" ,object nil 50)))

(prompt :prompt "test" :sources 'test-source)

```

the line `("Attribute-1" ,object nil)` should *not* be possible. As in typing, we want things to be precisely mentioned, and not inferred. Therefore, I suggest that it is a required argument of the column! That is, the user should *have* to write `("Attribute-1" ,object nil 1)`.

@Ambrevar I know you are a strong proponent of typing, do you agree? Also, @aartaka what do you think?

**Comment by Ambrevar** [comment]
Yes, the lack of good definition for the object attributes beyond the 2nd value is clumsy.

The reason behind this is that it's not part of the `prompter` library, but a Nyxt extension to it.

Suggestion: only add a third value which is either a plist or a class.  Example:

```lisp
(defmethod prompter:object-attributes ((object string) (source test-source))
  `(("Attribute-1" ,object)
    ("Attribute-2" ,object (:width 50))
    ("Attribute-3" ,other (:spinneret-thingy ))
```

By the way, what's the 3rd value for?  @aartaka I can't find the documentation.
The only instance I found is in `gtk.lisp`:

```lisp

(defmethod prompter:object-attributes ((color string) (source color-source))
  `(("Color"
     ,color
     ,(let ((spinneret:*html-style* :tree))
        (spinneret:with-html-string
          (:span :style (format nil "background-color: ~a; color: ~a; border: 0.1em solid ~a; border-radius: 0.1em"
                                color (contrasting-color color) (contrasting-color color))
                 color))))))

```

**Comment by aadcg** [comment]
@Ambrevar from the changelog:

```log
(:li "The third value in the " (:nxref :function 'prompter:object-attributes)
        " is interpreted as display HTML for the suggestion. See the color-picker
support as an example application for this feature.")
```

Also see this excerpt from `render-attributes` (the very last `loop`):

```lisp
(loop for (nil attribute attribute-display)
                               in (prompter:active-attributes suggestion :source source)
                             collect (:td :title attribute
                                          (if attribute-display
                                              (:raw attribute-display)
                                              (spinneret::escape-string attribute))))
```

**Comment by Ambrevar** [comment]
Thanks @aadcg.

This changelog entry should be backported into `prompt-buffer.lisp` I believe.
In any case, a dedicated class would resolve this lack of documentation.

**Comment by aartaka** [comment]
Or we can wrap around `prompter:object-attributes` with `nyxt:object-attributes` to enforce typing and clarify documentation!

**Comment by aartaka** [comment]
As for your remark, @jmercouris, the design of optional attribute values followed that of an existing Lispy pattern: `&optional` function arguments. They are positional, but we can omit them, and they have reasonable default values. It's somewhat fuzzy in how `&optional` args work, but they are useful nonetheless. For lots of use-cases, including display configuration.

The user-facing configuration mechanism should be fuzzy, DWIM-ish, and user-friendly, and optional character of the attributes beyond the second one fits this strategy well enough strategy.

You can add more type checks on top to make sure the values are valid HTML strings or numbers, but don't take away the optionality of it!

**Comment by jmercouris** [comment]
I agree, follow through with fewer wishlists and less todo items scattered throughout the codebase... now or never.

**Comment by jmercouris** [comment]
I believe I am against it, I would rather not mess with it now.
